File: RoomArchComponents/app/src/main/java/flatstack/com/roomarchcomponents/screens/BaseFragment.java
package flatstack.com.roomarchcomponents.screens;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import butterknife.ButterKnife;
import butterknife.Unbinder;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public abstract class BaseFragment extends Fragment {

    protected abstract ActivityInfo getActivityInfo();

    private CompositeDisposable subscriptions = new CompositeDisposable();
    private Unbinder bindings;

    public Disposable subscribe(Disposable disposable) {
        subscriptions.add(disposable);
        return disposable;
    }

    @Override public View onCreateView(LayoutInflater inflater,
                                       ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(getActivityInfo().getLayoutId(), container, false);
        bindings = ButterKnife.bind(this, v);
        return v;
    }

    @Override public void onDestroyView() {
        super.onDestroyView();
        bindings.unbind();
    }

    @Override public void onStop() {
        super.onStop();
        subscriptions.clear();
    }

}

File: kotlin-base/app/src/main/java/com/flatstack/android/utils/ui/BaseFragment.kt
package com.flatstack.android.utils.ui

import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.ButterKnife

abstract class BaseFragment : Fragment() {

    @get:LayoutRes abstract val layoutRes: Int

    override fun onCreate(savedState: Bundle?) {
        super.onCreate(savedState)
        if (savedState != null) {
            restoreState(savedState)
        }
    }

    fun restoreState(savedState: Bundle) {}

    override fun onCreateView(inflater: LayoutInflater?,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater!!.inflate(layoutRes, container, false)
        ButterKnife.bind(this, rootView)
        return rootView
    }
}
